pkg/retriever/directcandidatesource.go

```diff
@@ -15,20 +15,31 @@ import (
 	"github.com/libp2p/go-libp2p/core/peer"
 )
 
-var _ CandidateFinder = &DirectCandidateFinder{}
+var _ types.CandidateSource = &DirectCandidateSource{}
 
-// DirectCandidateFinder finds candidate protocols from a fixed set of peers
-type DirectCandidateFinder struct {
+// DirectCandidateSource finds candidate protocols from a fixed set of peers
+type DirectCandidateSource struct {
 	h         host.Host
-	providers []peer.AddrInfo
+	providers []types.Provider
 }
 
-// NewDirectCandidateFinder returns a new DirectCandidateFinder for the given providers
-func NewDirectCandidateFinder(h host.Host, providers []peer.AddrInfo) *DirectCandidateFinder {
-	return &DirectCandidateFinder{
-		h:         h,
+type Option func(*DirectCandidateSource)
+
+func WithLibp2pCandidateDiscovery(h host.Host) Option {
```

Comment:
```suggestion
// WithLibp2pCandidateDiscovery sets a libp2p Host for the DirectCandidateFinder.
// If a Host is set, the providers will be queried to discover available protocols, otherwise
// all protocols will be assumed by default.
func WithLibp2pCandidateDiscovery(h host.Host) Option {
```